### PR TITLE
CDAP-5186 implement aggregator plugin type

### PR DESCRIPTION
**Pull request by albertshau**
Implementing the aggregator plugin type for mapreduce engine.
Each aggregator translates to a reduce step in a mapreduce job.
The group by method is called in the mapper in order to decide
what to reduce on. The aggregate method is called first thing in
the reducer.

Moved most of the logic formerly in ETLMapper into TransformRunner
so that it can be re-used in the new added ETLReducer. Also moved
OutputWriter classes that were internal to ETLMapReduce out of the
class so that they can be re-used in the TransformRunner.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT3741

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  public StructuredRecord get() {
+    return record;
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  @Override
+  public void write(DataOutput out) throws IOException {
+    String schemaStr = record.getSchema().toString();
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: Can we open a JIRA for improving this. We are writing schema for all records.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 2]
```diff
@@ -190,9 +211,61 @@ public void beforeSubmit(MapReduceContext context) throws Exception {
...
+      }
+      if (outputValClass == null) {
+        outputValClass = TypeChecker.getGroupValueClass(aggregator);
+      }
+      hConf.set(GROUP_KEY_CLASS, outputKeyClass.getName());
+      hConf.set(GROUP_VAL_CLASS, outputValClass.getName());
+      // in case the classes are not a WritableComparable, but is some common type we support
+      // for example, a String or a StructuredRecord
+      WritableConversion writableConversion = WritableConversions.getConversion(outputKeyClass.getName());
+      if (writableConversion != null) {
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: Will we hit the case where writableConversion is null? Some class that is provided by the user that we don't support in WritableConversions?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/WritableConversions.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Functions to convert common classes to their WritableComparable equivalent and vice versa.
+ */
+public class WritableConversions {
+  // classname -> function for functions that transform common classes that should be transformed into classes usable
+  // as mapper keys or values.
+  private static final Map<String, WritableConversion<?, ?>> conversions =
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/WritableConversions.java]: all caps CONVERSIONS.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ErrorSInk.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * The first issue is that this uses avro directly, which means the app requires a specific avro version,
+ * and plugins will run into issues if they want to use a different avro (CDAP-3809).
+ *
+ * The second issue is that it assumes everything is a StructuredRecord and casts blindly.
+ *
+ * @param <KEY_OUT> output key type
+ * @param <VAL_OUT> output value type
+ */
+class ErrorSink<KEY_OUT, VAL_OUT> {
```
Comment (sreevatsanraman) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ErrorSInk.java]: nit: You might want to rename the file as "ErrorSink" and not "ErrorSInk"

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/WritableConversions.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    ImmutableMap.<String, WritableConversion<?, ?>>builder()
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/WritableConversions.java]: Constructing the map like this is a bit hard to read. I think it's better to use a static initializer block. Also, can the `WritableConversion` class have method like `toWritable` and `fromWritable` (or `wrap`, `unwrap`) directly instead of returning a function for doing so?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java (hunk 1 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: yep will do.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java (hunk 2 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: yeah, can add a comment

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/WritableConversions.java (hunk 5 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/WritableConversions.java]: yeah... dont know why i have a method that returns a function.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    String recordStr = StructuredRecordStringConverter.toJsonString(record);
+    out.writeInt(schemaStr.length());
+    out.write(Bytes.toBytes(schemaStr));
+    out.writeInt(recordStr.length());
+    out.write(Bytes.toBytes(recordStr));
+  }
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: Why are we serializing to JSON? That is very verbose and will result in many bytes to shuffle. The JSON essentially repeats the schema... Ideally we would serialize the schema only once, to the config (it is the same for all records, right?) and then use an AVRO-like encoding for each record.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/**
+ * Wrapper around a {@link StructuredRecord} so that it can be used as the output key and/or value of a mapper.
+ * This is not very performant and must be improved soon (CDAP-5347).
+ */
+public class StructuredRecordWritable implements WritableComparable<StructuredRecordWritable> {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: Why do we need a wrapper? Can we just make StructuredRecord implement writable?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.format.StructuredRecordStringConverter;
+import org.apache.hadoop.io.WritableComparable;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/**
+ * Wrapper around a {@link StructuredRecord} so that it can be used as the output key and/or value of a mapper.
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: Does it have to be comparable?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java (hunk 6 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: there is a jira open for this. If we do know the schema, it should not be written and we shouldn't be using json.  But there is no code I can readily use to do this, so was going to do it in a separate pr

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/IdentityConversion.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.apache.hadoop.io.Writable;
+
+/**
+ * Identity conversion that doesn't convert anything but just casts.
+ *
+ * @param <T> type of object to convert to a Writable
+ * @param <W> the Writable type to convert to
+ */
+public class IdentityConversion<T, W extends Writable> extends WritableConversion<T, W> {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/conversion/IdentityConversion.java]: if it's identity, then why does it need two type parameters? I would think it is:
```
public class IdentityConversion<T extends Writable> extends WritableConversion<T, T>
```

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 10]
```diff
@@ -110,11 +113,29 @@ public void configure() {
     setMapperResources(phaseSpec.getResources());
     setDriverResources(phaseSpec.getResources());
 
+    // These should never happen unless there is a bug in the planner. The planner is supposed to ensure this.
     if (phaseSpec.getPhase().getSources().size() != 1) {
-      throw new IllegalArgumentException("Pipeline must contain exactly one source.");
+      throw new IllegalArgumentException("Pipeline phase must contain exactly one source.");
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: In this case, it would be smart to include the phase and the two sources in the error message. Otherwise it will be hard to debug.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 11]
```diff
@@ -110,11 +113,29 @@ public void configure() {
...
     }
     if (phaseSpec.getPhase().getSinks().isEmpty()) {
-      throw new IllegalArgumentException("Pipeline must contain at least one sink.");
+      throw new IllegalArgumentException("Pipeline phase must contain at least one sink.");
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: similar here.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 12]
```diff
@@ -110,11 +113,29 @@ public void configure() {
...
+    }
+    Set<StageInfo> aggregators = phaseSpec.getPhase().getStagesOfType(BatchAggregator.PLUGIN_TYPE);
+    if (aggregators.size() > 1) {
+      throw new IllegalArgumentException("Pipeline phase cannot have more than one aggregator.");
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: and here.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 13]
```diff
@@ -110,11 +113,29 @@ public void configure() {
...
+    } else if (!aggregators.isEmpty()) {
+      String aggregatorName = aggregators.iterator().next().getName();
+      PipelinePhase mapperPipeline = phaseSpec.getPhase().subsetTo(ImmutableSet.of(aggregatorName));
+      for (StageInfo stageInfo : mapperPipeline) {
+        // error datasets are not supported in the map phase of a mapreduce, because we can only
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: how would that manifest itself in a pipeline? I am asking because users who specify a pipeline do not think in terms of stages and whether a map phase connects to an aggregator, but they may encounter this error message?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 14]
```diff
@@ -110,11 +113,29 @@ public void configure() {
...
+        // write out the group key and not error dataset data.
+        // we need to re-think how error datasets are done. perhaps they are just sinks instead of a special thing.
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: That sounds like a good idea. But would it mean that a pipeline can have more phases, only to collect the errors?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 15]
```diff
@@ -190,9 +211,62 @@ public void beforeSubmit(MapReduceContext context) throws Exception {
...
+      // if the conversion is null, it means the user is using their own object.
+      if (writableConversion != null) {
+        outputKeyClass = writableConversion.getWritableClass();
+      }
+      writableConversion = WritableConversions.getConversion(outputValClass.getName());
+      if (writableConversion != null) {
+        outputValClass = writableConversion.getWritableClass();
+      }
+      // check classes here instead of letting mapreduce do it, since mapreduce throws a cryptic error
+      if (!WritableComparable.class.isAssignableFrom(outputKeyClass)) {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: see my previous comment, does MR require that it is comparable?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 16]
```diff
@@ -317,94 +308,44 @@ public void map(Object key, Object value, Context context) throws IOException, I
...
-        errorMsg = String.format("Error Entry is of type %s, only a record of type %s is supported currently",
-                                 invalidEntry.getInvalidRecord().getClass().getName(),
-                                 StructuredRecord.class.getName());
+    @Override
+    protected void reduce(Object key, Iterable values, Context context) throws IOException, InterruptedException {
+      try {
+        transformRunner.transform(key, values.iterator());
+      } catch (Exception e) {
+        LOG.error("Exception thrown in BatchDriver Reducer.", e);
+        Throwables.propagate(e);
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: why log and throw? That way it gets logged twice?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ErrorSInk.java (hunk 4 above)
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ErrorSInk.java]: Is this an ETL sink? If not, can it be named differently?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformExecutorFactory.java [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    private final WritableConversion<GROUP_KEY, OUT_KEY> keyConversion;
+    private final WritableConversion<GROUP_VAL, OUT_VAL> valConversion;
+
+    public AggregatorTransformation(Aggregator<GROUP_KEY, GROUP_VAL, ?> aggregator,
+                                    StageMetrics stageMetrics,
+                                    String groupKeyClassName,
+                                    String groupValClassName) {
+      this.aggregator = aggregator;
+      this.stageMetrics = stageMetrics;
+      WritableConversion<GROUP_KEY, OUT_KEY> keyConversion = WritableConversions.getConversion(groupKeyClassName);
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformExecutorFactory.java]: can this return null? It does here: https://github.com/caskdata/cdap/pull/5341/files#diff-f7cf8103359905eb8aeacfc3074876b0R244

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformExecutorFactory.java [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+      WritableConversion<GROUP_VAL, OUT_VAL> valConversion = WritableConversions.getConversion(groupValClassName);
+      this.keyConversion = keyConversion == null ? new IdentityConversion<GROUP_KEY, OUT_KEY>() : keyConversion;
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformExecutorFactory.java]: so if the user uses his own class, then it must be writable/comparable already. Ok.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformRunner.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    Map<String, Map<String, String>> runtimeArgs,
+    PipelinePhase pipelinePhase) throws Exception {
+
+    TransformExecutorFactory<KeyValue<Object, Object>> transformExecutorFactory =
+      new MapperTransformExecutorFactory<>(context, pluginInstantiator, metrics, runtimeArgs);
+
+    Set<StageInfo> aggregators = pipelinePhase.getStagesOfType(BatchAggregator.PLUGIN_TYPE);
+    if (!aggregators.isEmpty()) {
+      String aggregatorName = aggregators.iterator().next().getName();
+      return transformExecutorFactory.create(pipelinePhase.subsetTo(ImmutableSet.of(aggregatorName)));
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformRunner.java]: feels like I have seen this before in this PR. Is it code duplication?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ReducerTransformExecutorFactory.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+import java.util.Iterator;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Creates transform executors for mapreduce programs.
+ *
+ * @param <T> the type of input for the created transform executors
+ */
+public class ReducerTransformExecutorFactory<T> extends TransformExecutorFactory<T> {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ReducerTransformExecutorFactory.java]: this class appears to have a lot of the same code as MapperTransformExecutorFactory. Can they be refactored into one class?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+    try {
+      KeyValue<KEY, VALUE> input = new KeyValue<>(key, value);
+      TransformResponse transformResponse = transformExecutor.runOneIteration(input);
+      for (Map.Entry<String, Collection<Object>> transformedEntry : transformResponse.getSinksResults().entrySet()) {
+        for (Object transformedRecord : transformedEntry.getValue()) {
+          outputWriter.write(transformedEntry.getKey(), (KeyValue<Object, Object>) transformedRecord);
+        }
+      }
+
+      for (Map.Entry<String, Collection<InvalidEntry<Object>>> errorEntries :
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: call this errorEntry?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+        transformResponse.getMapTransformIdToErrorEmitter().entrySet()) {
+
+        if (transformsWithoutErrorDataset.contains(errorEntries.getKey())) {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: is this equivalent to errorEntry.getValue() == null? Or could a transform that has no error dataset emit an error record?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+          continue;
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: when can this happen? If the same transform appears multiple times?

Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: oh is this to avoid logging the same error many times?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+                     errorEntries.getKey());
+            transformsWithoutErrorDataset.add(errorEntries.getKey());
+          } else {
+            transformErrorSinkMap.get(errorEntries.getKey()).write(errorEntries.getValue());
+          }
+        }
+      }
+      transformExecutor.resetEmitter();
+    } catch (Exception e) {
+      LOG.error("Exception thrown in BatchDriver Mapper.", e);
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: log and rethrow?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java [hunk 25]
```diff
@@ -48,4 +48,5 @@ public SparkTransformExecutorFactory(PluginContext pluginContext,
   protected BatchRuntimeContext createRuntimeContext(String stageName) {
     return new SparkBatchRuntimeContext(pluginContext, metrics, logicalStartTime, runtimeArgs, stageName);
   }
+
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java]: adding a new line?

Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java]: Do we support groub-bys and aggregations with spark?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java [hunk 26]
```diff
@@ -62,21 +59,58 @@ private PipelinePhase(Map<String, Set<StageInfo>> stages, Map<String, Set<String
...
+   */
+  public PipelinePhase subsetFrom(Set<String> newSources) {
+    return getSubset(dag.subsetFrom(newSources));
+  }
+
+  private PipelinePhase getSubset(Dag subsetDag) {
+    Map<String, Set<StageInfo>> subsetStages = new HashMap<>();
+    for (Map.Entry<String, Set<StageInfo>> stagesEntry : stages.entrySet()) {
+      Set<StageInfo> stagesOfType = new HashSet<>();
+      for (StageInfo stageInfo : stagesEntry.getValue()) {
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java]: use Sets.filter()?

File: cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/test/java/co/cask/cdap/etl/datapipeline/mock/IdentityAggregator.java [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+import co.cask.cdap.etl.api.StageConfigurer;
+import co.cask.cdap.etl.api.batch.BatchAggregator;
+import co.cask.cdap.etl.proto.v2.ETLPlugin;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Identity aggregation for testing.
```
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-data-pipeline/src/test/java/co/cask/cdap/etl/datapipeline/mock/IdentityAggregator.java]: it's not identity. It picks a random value from all values with the same key.

**Comment by anew** [comment]
@ashau I did one pass over this PR. please see comments.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java (hunk 6 above)
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: that's ok

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java (hunk 7 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/StructuredRecordWritable.java]: Ideally we could, but since StructuredRecord is in cdap-api, it would mean adding hadoop as a dependency of cdap-api

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java (hunk 13 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: perhaps the terminology is confusing here, but a 'stage' is a node in the pipeline and has a unique name, so at least they should be able to identify the problem node.  Why it is a problem will be very confusing though, as it depends on where it sits in the pipeline.... I'm tempted to just remove support altogether

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java (hunk 14 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: yeah if that is how its done, it could add more phases.  Not sure if thats what we should do

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java (hunk 16 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: shouldn't be logging here, will remove

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java (hunk 23 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: im not entirely sure... most of this class is just whatever was being previously done in the mapper.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java (hunk 24 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: will remove.  This method should just declare that it throws Exception

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java (hunk 23 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: ok this is if the transform emitted errors, but didn't configure an error dataset to write to.  I will add a comment. We need to revisit error datasets...

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java]: er yeah actually now i get it, it's like you said, used to avoid logging the same error multiple times.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java (hunk 25 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java]: The data-pipeline app doesn't support spark yet, still needs to be implemented

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformRunner.java (hunk 19 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformRunner.java]: its similar, but not the same, as the ReducerTransformRunner, except the reducer gets the part of the pipeline starting at the aggregator, whereas here we get the part of the pipeline ending at the aggregator

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ReducerTransformExecutorFactory.java (hunk 20 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ReducerTransformExecutorFactory.java]: should be able to. will take another pass to clean things up.

**Comment by albertshau** [comment]
thanks for the review @anew. I have addressed comments, could you take another look?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java (hunk 13 above)
Comment (anew) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java]: You are tempted to remove support for error datasets? Let's keep that for another PR - if at all.

**Comment by anew** [comment]
Thanks for addressing all comments. LGTM :+1:

**Comment by sreevatsanraman** [comment]
There was an unrelated build failure due to CLIMain test failing. 
Filed a JIRA for that: https://issues.cask.co/browse/CDAP-5370
Re-running build
